hs-bindgen/src/HsBindgen/Hs/AST.hs

```diff
@@ -115,6 +124,58 @@ data DataDecl n f = MkDataDecl
   deriving stock (GHC.Generic)
   deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
 
+-- | Variable or function declaration.
+type VarDecl :: PHOAS
+data VarDecl f =
+  VarDecl
+    -- | Name of variable/function.
+    { varDeclName :: HsName NsVar
+    -- | Type of variable/function.
+    , varDeclType :: VarDeclType f
+    -- | RHS of variable/function.
+    , varDeclBody :: VarDeclRHS f
+    }
+  deriving stock (GHC.Generic)
+  deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
+
+type VarDeclType :: PHOAS
+data VarDeclType f
+  = TyLamTy (HsName NsTypeVar) (VarDeclType f)
```

Comment:
This is wrong. We have PHOAS, but you ahve explicit name for local binding. This is combining worst of both.

Reply:
And also this is forall-binder, not type-level lambda. Is it?

Reply:
Thanks, fixed now. However, I wasn't sure how to deal with macros referring to other macros; ideally we should be able to resolve them directly but currently in the backends I just use plain names (e.g. in the TH backend I use `mkName`).

Reply:
> macros referring to other macros; 

We simply cannot have "fresh" names for top-level definitions. But that's not a real problem, as top-level names are part of API, so they cannot  really be fresh anyway.